evm/vm/base.py

```diff
@@ -132,6 +133,7 @@ def execute_bytecode(self,
     # Mining
     #
     def import_block(self, block):
+        self.receipts = []
```

Comment:
This stood out to me.  What if instead of tracking this state we instead returned the receipt from `apply_transaction` and accumulated the receipts in the loop [here](https://github.com/ethereum/py-evm/pull/565/files#diff-7a66be18d19df3f35fdebb216454a46fR149)

It looks like this change goes well with the changes within the `VMState` class.  It looks like we could remove the need for that class to be passed the receipts since I *think* this changes things so that the only thing it needs to know is the `gas_used` value from the latest receipt.

Reply:
Yeah, but should we then pass `gas_used` as an argument to `__init__()` and keep updating it in `add_transaction()`? Or did you have something else in mind?

Reply:
Honestly not sure.  Nothing specific in mind for exact implementation but starting with `gas_used` being passed into `__init__` seems like a good start.  It'd be nice if it didn't need to be mutated but i'm more ok with mutating that value than the receipt storage on the `VM` instance.

Reply:
So, we need `gas_used` in `BaseState` so I'm passing it via `__init__()` and updating it on `BaseState.add_transaction()`, but in `BaseVM` we also need that value in the `.state` @property, which is currently looking that up from `BaseVM.receipts` but we could instead have a `BaseVM.gas_used` and update that on `BaseVM.apply_transaction()`. Would you prefer that?

Reply:
That sounds viable.  We can continue to iterate on this to I'd recommend just picking something that seems good enough.

Reply:
We actually need to keep the receipts somewhere in order to generate the receipt trie